File: tag/th/tag_char_rnn.lua
require 'torch'
require 'nn'
require 'optim'
require 'utils'
require 'torchure'
require 'train'
require 'data'
require 'cudnn'
require 'model'


DEF_BATCHSZ = 50
DEF_TSF = './data/twpos-data-v0.3/oct27.splits/oct27.train'
DEF_VSF = './data/twpos-data-v0.3/oct27.splits/oct27.dev'
DEF_ESF = './data/twpos-data-v0.3/oct27.splits/oct27.test'
DEF_FILE_OUT = 'rnn-tagger.model'
DEF_EVAL_OUT = 'rnn-tagger-test.txt'
DEF_PATIENCE = 10
DEF_RNN = 'blstm'
DEF_NUM_RNN = 1
DEF_OPTIM = 'adadelta'
DEF_EPOCHS = 60
DEF_ETA = 0.32
DEF_CFILTSZ = '{1,2,3,4,5,7}'
DEF_HSZ = 100
DEF_CHARSZ = 16
DEF_WSZ = 50
DEF_PROC = 'gpu'
DEF_CLIP = 5
DEF_DECAY = 1e-7
DEF_MOM = 0.0
DEF_UNIF = 0.25
DEF_PDROP = 0.5
DEF_MXLEN = 40
DEF_VALSPLIT = 0.15
DEF_OUT_OF_CORE = false
DEF_EMBED = 'NONE'
DEF_CEMBED = 'NONE'
torch.setdefaulttensortype('torch.FloatTensor')

--------------------------
-- Command line handling
--------------------------
local cmd = torch.CmdLine()
cmd:option('-cembed', DEF_CEMBED, 'Character-level pre-trained embeddings')
cmd:option('-embed', DEF_EMBED, 'Word-level pre-trained embeddings')
cmd:option('-batchsz', DEF_BATCHSZ, 'Batch size')
cmd:option('-save', DEF_FILE_OUT, 'Save model to')
cmd:option('-conll_output', DEF_EVAL_OUT, 'Place to put test CONLL file')
cmd:option('-rnn', DEF_RNN, 'RNN type (blstm|lstm) default is blstm')
cmd:option('-numrnn', DEF_NUM_RNN, 'The depth of stacked RNNs')
cmd:option('-train', DEF_TSF, 'Training file')
cmd:option('-valid', DEF_VSF, 'Validation file (optional)')
cmd:option('-eval', DEF_ESF, 'Testing file')
cmd:option('-valsplit', DEF_VALSPLIT, 'Fraction training used for validation if no set is given')
cmd:option('-unif', DEF_UNIF, 'Vocab embedding initialization for non-attested attributes')
cmd:option('-optim', DEF_OPTIM, 'Optimization methodx (sgd|adagrad|adadelta|adam)')
cmd:option('-epochs', DEF_EPOCHS)
cmd:option('-eta', DEF_ETA)
cmd:option('-clip', DEF_CLIP)
cmd:option('-decay', DEF_DECAY)
cmd:option('-mom', DEF_MOM, 'Momentum for SGD')
cmd:option('-cfiltsz', DEF_CFILTSZ, 'Convolution filter width')
cmd:option('-hsz', DEF_HSZ, 'Hidden layer units')
cmd:option('-charsz', DEF_CHARSZ, 'Character embedding depth')
cmd:option('-wsz', DEF_WSZ, 'Word embedding depth')
cmd:option('-proc', DEF_PROC)
cmd:option('-patience', DEF_PATIENCE)
cmd:option('-dropout', DEF_PDROP, 'Dropout probability')
cmd:option('-ooc', DEF_OUT_OF_CORE, 'Should data batches be file-backed?')
cmd:option('-mxlen', DEF_MXLEN, 'Max sentence length')
cmd:option('-cbow', false, 'Do CBOW for characters')
local opt = cmd:parse(arg)
if opt.cbow then
   opt.cfiltsz = "{1}"
end
opt.cfiltsz = loadstring("return " .. opt.cfiltsz)()

----------------------------------------
-- Optimization
----------------------------------------

state, optmeth = optimMethod(opt)

--------------------------------------
-- Processing on GPU or CPU
----------------------------------------
opt.gpu = false
if opt.proc == 'gpu' then
   opt.gpu = true
   require 'cutorch'
   require 'cunn'
else
   opt.proc = 'cpu'
end
print('Processing on ' .. opt.proc)

require 'nnx'
require 'rnn'

local vocab = nil

maxw, vocab_ch, vocab_word = conllBuildVocab({opt.train, opt.eval, opt.valid})

maxw = math.min(maxw, opt.mxlen)
print('Max word length ' .. maxw)
---------------------------------------
-- Load Word2Vec Model(s)
---------------------------------------
local f2i = {}
local finetune = true

if opt.embed ~= 'NONE' then
   print('Loading word embeddings ' .. opt.embed)
   word_vec = Word2VecLookupTable(opt.embed, vocab_word, opt.unif, false, finetune)
else
   word_vec = {}
   word_vec.dsz = 0
end

-- Character embeddings
if opt.cembed == 'NONE' then
   if opt.charsz ~= opt.wsz and opt.cbow == true then
      print('Warning, you have opted for CBOW char embeddings, but have provided differing sizes for char embedding depth and word depth.  This is not possible, forcing char embedding depth to be word depth ' .. opt.wsz)
      opt.charsz = opt.wsz
   end

   wch_vec = VocabLookupTable(vocab_ch, opt.charsz, opt.unif)
else
   print('Using pre-trained character embeddings ' .. opt.cembed)
   wch_vec = Word2VecLookupTable(opt.cembed, vocab_ch, opt.unif, false, finetune)
   opt.charsz = wch_vec.dsz
   if opt.charsz ~= opt.wsz and opt.cbow == true then
      print('Warning, you have opted for CBOW char embeddings, and have provided pre-trained char vector embeddings.  To make this work, setting word vector size to character vector size ' .. opt.charsz)
      opt.wsz = opt.charsz
   end

end

--print(word_vec.vocab["<PADDING>"])
function afterhook() 
   wch_vec.weight[1]:zero()
   if word_vec.weight then
      word_vec.weight[1]:zero()
   end
end
opt.afteroptim = afterhook

-- Load Feature Vectors
---------------------------------------
ts,f2i,_ = conllSentsToIndices(opt.train, word_vec, wch_vec, maxw, f2i, opt)
print('Loaded training data')

if opt.valid ~= 'NONE' then
   print('Using provided validation data')
   vs,f2i,_ = conllSentsToIndices(opt.valid, word_vec, wch_vec, maxw, f2i, opt)
else
   ts,vs = validSplit(ts, opt.valsplit, opt.ooc)
   print('Created validation split')
end

es,f2i,txts = conllSentsToIndices(opt.eval, word_vec, wch_vec, maxw, f2i, opt)
-- print(txts)
print(f2i)
print('Using ' .. ts:size() .. ' examples for training')
print('Using ' .. vs:size() .. ' examples for validation')
print('Using ' .. es:size() .. ' examples for test')

local i2f = revlut(f2i)
local nc = #i2f
print('Number of classes ' .. nc)
---------------------------------------
-- Build model and criterion
---------------------------------------
local crit = createTaggerCrit(opt.gpu)


local model = createTaggerModel(word_vec, wch_vec, opt, nc)

local maxacc = 0
local lastImproved = 0

for i=1,opt.epochs do
    print('Training epoch ' .. i)
    
    trainTaggerEpoch(crit, model, ts, optmeth, opt)
    local loss, acc = testTagger('Validation', model, vs, crit, i2f, opt)

    if acc > maxacc then
       lastImproved = i
       maxacc = acc
       print('Highest dev acc achieved yet -- writing model')
       saveModel(model, opt.save, opt.gpu)
    end
    if (i - lastImproved) > opt.patience then
       print('Stopping due to persistent failures to improve')
       break
    end
end

print("-----------------------------------------------------")
print('Highest dev acc: ' .. maxacc)
print('=====================================================')
print('Evaluating best model on test data')

model = loadModel(opt.save, opt.gpu)
print('Reloaded best model')

-- For final evaluation, dont batch, want to get all examples once
-- Also this will write out each example to the opt.output file, which can
-- be used with conlleval.pl to test overall performance
opt.batchsz = 1
local loss, acc = testTagger('Test', model, es, crit, i2f, opt, txts)

print("-----------------------------------------------------")
print('Test loss '.. loss)
print('Test acc ' .. acc)
print('=====================================================')

File: tag/tf/tag_char_rnn.py
import tensorflow as tf
import numpy as np
import math
import w2v
from data import conllBuildVocab
from data import conllSentsToIndices
from data import batch
from data import revlut
from data import validSplit
from model import TaggerModel, vizEmbeddings
from train import Trainer
import time

DEF_BATCHSZ = 50
DEF_TSF = './data/twpos-data-v0.3/oct27.splits/oct27.train'
DEF_VSF = './data/twpos-data-v0.3/oct27.splits/oct27.dev'
DEF_ESF = './data/twpos-data-v0.3/oct27.splits/oct27.test'
DEF_FILE_OUT = 'rnn-tagger.model'
DEF_EVAL_OUT = 'rnn-tagger-test.txt'
DEF_PATIENCE = 10
DEF_RNN = 'blstm'
DEF_NUM_RNN = 1
DEF_OPTIM = 'adadelta'
DEF_EPOCHS = 60
DEF_ETA = 0.32
DEF_CFILTSZ = '1,2,3,4,5,7'
DEF_HSZ = 100
DEF_CHARSZ = 16
DEF_WSZ = 50
DEF_PROC = 'gpu'
DEF_CLIP = 5
DEF_DECAY = 1e-7
DEF_MOM = 0.0
DEF_UNIF = 0.25
DEF_PDROP = 0.5
DEF_MXLEN = 40
DEF_VALSPLIT = 0.15
DEF_EMBED = None
DEF_CEMBED = None

flags = tf.app.flags
FLAGS = flags.FLAGS
flags.DEFINE_float('eta', DEF_ETA, 'Initial learning rate.')
flags.DEFINE_string('embed', DEF_EMBED, 'Word2Vec embeddings file')
flags.DEFINE_string('cembed', DEF_CEMBED, 'Word2Vec char embeddings file')
flags.DEFINE_string('optim', DEF_OPTIM, 'Optim method')
flags.DEFINE_float('dropout', DEF_PDROP, 'Dropout probability')
flags.DEFINE_string('train', DEF_TSF, 'Training file')
flags.DEFINE_string('valid', DEF_VSF, 'Validation file')
flags.DEFINE_string('test', DEF_ESF, 'Test file')
flags.DEFINE_string('rnn', DEF_RNN, 'RNN type')
flags.DEFINE_integer('numrnn', DEF_NUM_RNN, 'The depth of stacked RNNs')
flags.DEFINE_string('outdir', 'out', 'Directory to put the output')
flags.DEFINE_float('unif', DEF_UNIF, 'Initializer bounds for embeddings')
flags.DEFINE_float('clip', DEF_CLIP, 'Gradient clipping cutoff')
flags.DEFINE_integer('epochs', DEF_EPOCHS, 'Number of epochs')
flags.DEFINE_integer('batchsz', DEF_BATCHSZ, 'Batch size')
flags.DEFINE_integer('mxlen', DEF_MXLEN, 'Max length')
flags.DEFINE_string('cfiltsz', DEF_CFILTSZ, 'Character filter sizes')
#flags.DEFINE_integer('charsz', 150, 'Char embedding depth')
flags.DEFINE_integer('charsz', DEF_CHARSZ, 'Char embedding depth')
flags.DEFINE_integer('patience', DEF_PATIENCE, 'Patience')
flags.DEFINE_integer('hsz', DEF_HSZ, 'Hidden layer size')
flags.DEFINE_integer('wsz', DEF_WSZ, 'Word embedding depth')
flags.DEFINE_float('valsplit', DEF_VALSPLIT, 'Validation split if no valid set')
#flags.DEFINE_string('cfiltsz', '0', 'Character filter sizes')
flags.DEFINE_boolean('cbow', False, 'Do CBOW for characters')
flags.DEFINE_string('save', DEF_FILE_OUT, 'Save basename')
#flags.DEFINE_boolean('crf', False, 'Use CRF on top')



maxw, vocab_ch, vocab_word = conllBuildVocab([FLAGS.train, 
                                              FLAGS.test, 
                                              FLAGS.valid])

maxw = min(maxw, FLAGS.mxlen)
print('Max word length %d' % maxw)


if FLAGS.cbow is True:
    print('Using CBOW char embeddings')
    FLAGS.cfiltsz = '0'
else:
    print('Using convolutional char embeddings')
word_vec = None
if FLAGS.embed:
    word_vec = w2v.Word2VecModel(FLAGS.embed, vocab_word, FLAGS.unif)

if FLAGS.cembed:
    print('Using pre-trained character embeddings ' + FLAGS.cembed)
    char_vec = w2v.Word2VecModel(FLAGS.cembed, vocab_ch, FLAGS.unif)
    FLAGS.charsz = char_vec.dsz
    if FLAGS.charsz != FLAGS.wsz and FLAGS.cbow is True:
        print('Warning, you have opted for CBOW char embeddings, and have provided pre-trained char vector embeddings.  To make this work, setting word vector size to character vector size ' + FLAGS.charsz)
        opt.wsz = opt.charsz
else:
    if FLAGS.charsz != FLAGS.wsz and FLAGS.cbow is True:
        print('Warning, you have opted for CBOW char embeddings, but have provided differing sizes for char embedding depth and word depth.  This is not possible, forcing char embedding depth to be word depth ' + FLAGS.wsz)
        FLAGS.charsz = FLAGS.wsz

    char_vec = w2v.RandomInitVecModel(FLAGS.charsz, vocab_ch, FLAGS.unif)


f2i = {"<PAD>":0}

opts = { 'batchsz': FLAGS.batchsz,
         'mxlen': FLAGS.mxlen }

ts, f2i, _ = conllSentsToIndices(FLAGS.train, word_vec, char_vec, maxw, f2i, opts)
print('Loaded  training data')

if FLAGS.valid is not None:
    print('Using provided validation data')
    vs, f2i,_ = conllSentsToIndices(FLAGS.valid, word_vec, char_vec, maxw, f2i, opts)
else:
    ts, vs = validSplit(ts, FLAGS.valsplit)
    print('Created validation split')


es, f2i,txts = conllSentsToIndices(FLAGS.test, word_vec, char_vec, maxw, f2i, opts)
print('Loaded test data')

i2f = revlut(f2i)
print(i2f)
print('Using %d examples for training' % len(ts))
print('Using %d examples for validation' % len(vs))
print('Using %d examples for test' % len(es))


with tf.Graph().as_default():
    sess = tf.Session()
    with sess.as_default():

        model = TaggerModel()
        model.params(f2i,
                     word_vec,
                     char_vec,
                     FLAGS.mxlen,
                     maxw,
                     FLAGS.rnn,
                     FLAGS.wsz,
                     FLAGS.hsz,
                     FLAGS.cfiltsz)


        trainer = Trainer(sess, model, FLAGS.outdir, FLAGS.optim, FLAGS.eta)

        train_writer = trainer.writer()

        vizEmbeddings(char_vec, word_vec, FLAGS.outdir, train_writer)

        init = tf.global_variables_initializer()
        sess.run(init)
        trainer.prepare(tf.train.Saver())

        max_acc = 0
        last_improved = 0
        for i in range(FLAGS.epochs):
            print('Training epoch %d' % (i+1))
            trainer.train(ts, FLAGS.dropout, FLAGS.batchsz)
            avg_loss, this_acc = trainer.test(vs, FLAGS.batchsz, 'Validation')
            if this_acc > max_acc:
                max_acc = this_acc
                last_improved = i
                print('Highest dev acc achieved yet -- writing model')
                trainer.checkpoint(FLAGS.save)
            if (i - last_improved) > FLAGS.patience:
                print('Stopping due to persistent failures to improve')
                break


        print("-----------------------------------------------------")
        print('Highest dev acc %.2f' % (max_acc * 100.))
        print('=====================================================')
        print('Evaluating best model on test data')
        print('=====================================================')
        trainer.recover_last_checkpoint()
        #best_model = tf.train.latest_checkpoint(FLAGS.outdir + "/train/")
        #print("Reloading " + best_model)
        #saver.restore(sess, best_model)
        avg_loss, this_acc = trainer.test(es, 1, 'Test')
        print("-----------------------------------------------------")
        print('Test loss %.2f' % (avg_loss))
        print('Test acc %.2f' % (this_acc * 100.))
        print('=====================================================')
        # Write out model, graph and saver for future inference
        model.save(sess, FLAGS.outdir, FLAGS.save)
